src/config.rs

```diff
@@ -0,0 +1,122 @@
+use clap::Parser;
+use num_rational::Ratio;
+use serde::Serialize;
+
+use crate::seat::Seat;
+
+#[derive(Parser, Serialize, Debug)]
+pub struct Config {
+    #[arg(long)]
+    pub num_blocks: u64,
+    #[arg(long)]
+    pub num_shards: u64,
+    /// TODO mention excess validators are ignored/unassigned.
+    #[arg(long)]
+    pub seats_per_shard: u64,
+    /// The amount of stake required to get one seat.
+    #[arg(long)]
+    pub stake_per_seat: u64,
+    /// If the ratio of malicious stake is higher than this threshold, the shard is considered
+    /// corrupted, i.e. a security failure occured.
+    #[arg(long)]
+    pub max_malicious_stake_per_shard: Ratio<u64>,
+}
+
+impl Config {
+    #[cfg(test)]
+    pub fn new_mock() -> Self {
+        Self {
+            num_blocks: 1_000,
+            num_shards: 4,
+            seats_per_shard: 2,
+            stake_per_seat: 100,
+            max_malicious_stake_per_shard: Ratio::new(1, 3),
+        }
+    }
+
+    /// Returns the number of (full) seats that can be claimed by `stake`.
+    pub fn seats_per_stake(&self, stake: u64) -> u64 {
+        // Integer division in Rust returns the floor as described here
+        // https://doc.rust-lang.org/std/primitive.u64.html#method.div_euclid
+        stake / self.stake_per_seat
+    }
+
+    /// Returns the amount of seats for all shards that must be filled by validators.
+    pub fn total_seats(&self) -> u64 {
+        self.num_shards
+            .checked_mul(self.seats_per_shard)
+            .expect("min_required_seats should fit into return type")
+    }
+
+    // Collects the number of seats required for a shard by collecting `seats[shard_idx * i]` for `i
+    // in 0..n` where `n` is the number of seats per shard.
+    pub fn collect_seats_for_shard<'seats>(
```

Comment:
Why do we assume the seats are interleaved like this? If there are 3 shards then this assumes the seats are arranged as `[seat for 0, seat for 1, seat for 2, seat for 0, ...]`. But if the seats are given as a random shuffle anyway then we could equally well assume they are arranged as `[seat for 0, seat for 0, ..., seat for 1, seat for 1, ..., seat for 2, seat for 2, ...]`. Then This collection function is much simpler:

```rust
let start = start_idx * seats_per_shard;
seats[start..(start + seats_per_shard)]
```

Have I missed something about why the interleaving is important?

Reply:
If I understood it correctly, this interleaving corresponds to the the assignment of validators to shards described in the third bullet point [here](https://docs.google.com/document/d/1C-w4FNeXl8ZMd_Z_YxOf30XA1JM6eMDp5Nf3N-zzNWU/edit#heading=h.7wdbrn8ypjiw). I assumed this allows for leaner data structures resp. optimizations when implementing the algorithm in nearcore. I would suggest we ask for the motivation in the upcoming sync meeting.

There should have been a comment linking to above doc.I’ll add one.

Reply:
Agreed, makes sense to keep it simple given that it does not impact our results and for the simulation there are benefits or optimizations related to the more complicated interleaving. Will change it to the simpler assignment you proposed above.

Reply:
Changed in [66bdd69](https://github.com/mooori/sim-validator-assignment/pull/4/commits/66bdd69f24f519ca7ce7fbb98027772c39207abe).